src/negate.c

```diff
@@ -0,0 +1,14 @@
+#include "data.table.h"
+
+SEXP negate(SEXP x) {
+  if(TYPEOF(x) != LGLSXP) {
+    error("not logical or integer vector");
+  }
+  int n = length(x);
+  int *ansd = INTEGER(x);
+  for(int i=0;i<n;i++) {
+    ansd[i] = LOGICAL(x)[i] == TRUE?0:1;
```

Comment:
I like this approach to check `NA_LOGICAL`, so I've switched to using your first suggestion. I think it would be nice to be branchless, but I'm having a hard time understanding how the second approach would work. If the element is `NA_INTEGER` we would be setting the element of `ansd` to `NA_INTEGER xor true`, which doesn't intuitively have a value to me, but perhaps it can be evaluated?

Reply:
Great. I like the `match() == 0` approach you've done in the latest commit so the negate isn't needed on that result. Minor improvement would be to use `0L` to avoid the `0` double to integer coercion.
On this branchless question, if the element is NA_INTEGER,  `!= NA_INTEGER` will be false (not true) so that `NA_INTEGER xor false` will do nothing and leave `NA_INTEGER` as `NA_INTEGER`. And now I realize I meant NA_LOGICAL of course (not NA_INTEGER). Even though they're the same currently in R I maintain hope that one day type logical will be changed to be a single byte with NA_LOGICAL then having the value -127. Best be tidy and future proof where we can and use LOGICAL here because that is the correct type here.